src/plugins/gui/providers/ioniaProvider.ts

```diff
@@ -392,34 +413,75 @@ export const makeIoniaProvider: FiatProviderFactory<IoniaMethods> = {
+        const rateAmount = await getCardPurchaseRateAmount(quoteParams.displayCurrencyCode, RATE_QUOTE_CARD_AMOUNT)
+        const rateExchangeAmount = await quoteParams.wallet.nativeToDenomination(toBigNumberString(rateAmount), quoteParams.displayCurrencyCode)
+        console.log(`Ionia rates a $${RATE_QUOTE_CARD_AMOUNT} card at ${rateExchangeAmount} ${quoteParams.displayCurrencyCode}`)
 
         const price = RATE_QUOTE_CARD_AMOUNT / parseFloat(rateExchangeAmount)
         const cryptoAmount =
-          params.amountType === 'crypto' ? params.exchangeAmount : div(params.exchangeAmount, toBigNumberString(price), HARD_CURRENCY_PRECISION)
-        const fiatAmount = params.amountType === 'fiat' ? params.exchangeAmount : mul(params.exchangeAmount, toBigNumberString(price))
+          quoteParams.amountType === 'crypto' ? quoteParams.exchangeAmount : div(quoteParams.exchangeAmount, toBigNumberString(price), HARD_CURRENCY_PRECISION)
+        const fiatAmount = quoteParams.amountType === 'fiat' ? quoteParams.exchangeAmount : mul(quoteParams.exchangeAmount, toBigNumberString(price))
+
+        // Concurrently get the purchase card
+        const purchaseCardPromise = getPurchaseCard(quoteParams.displayCurrencyCode, parseFloat(fiatAmount))
 
         checkAmountMinMax(parseFloat(fiatAmount))
 
         const fiatProviderQuote: FiatProviderQuote = {
           providerId: makeIoniaProvider.providerId,
           partnerIcon: fiatProvider.partnerIcon,
           pluginDisplayName: '',
-          displayCurrencyCode: params.displayCurrencyCode,
+          displayCurrencyCode: quoteParams.displayCurrencyCode,
           cryptoAmount,
           isEstimate: true,
-          fiatCurrencyCode: params.fiatCurrencyCode,
+          fiatCurrencyCode: quoteParams.fiatCurrencyCode,
           fiatAmount,
           direction: 'sell',
-          regionCode: params.regionCode,
-          paymentTypes: params.paymentTypes,
-
-          approveQuote: async params => {},
+          regionCode: quoteParams.regionCode,
+          paymentTypes: quoteParams.paymentTypes,
+
+          approveQuote: async approveParams => {
+            const { showUi, coreWallet: wallet } = approveParams
+
+            // Unbox the latest purchase card:
+            const purchaseCard = await purchaseCardPromise
+
+            // Use the purchase card URI to complete the payment over the Payment Protocol:
+            const parsedUri: EdgeParsedUri & { paymentProtocolUrl?: string } = await wallet.parseUri(purchaseCard.uri, quoteParams.displayCurrencyCode)
+            const { paymentProtocolUrl } = parsedUri
+            if (paymentProtocolUrl == null) {
+              throw new Error(lstrings.missing_provider_payment_address_message)
+            }
+
+            // Check if the URI requires a warning to the user:
+            const approved = await addressWarnings(parsedUri, quoteParams.displayCurrencyCode)
```

Comment:
This breaks the plugin model. addressWarnings should be wrapped by `showUi`. Maybe in the cleanup PR